src/macaron/dependency_analyzer/cyclonedx.py

```diff
@@ -0,0 +1,206 @@
+
+    """
+    # TODO: use the official CycloneDX library to validate and deserialize BOM files
+    # once merged: https://github.com/CycloneDX/cyclonedx-python-lib/pull/290
+    if not os.path.exists(file_path):
+        raise CycloneDXParserError(f"Unable to locate the BOM file: {str(file_path)}.")
+
+    with open(file_path, encoding="utf8") as file:
+        try:
+            return dict(json.load(file))
+        except (ValueError, json.JSONDecodeError) as error:
+            raise CycloneDXParserError(f"Could not process the dependencies at {file_path}: {error}") from None
+
+
+def get_root_component(root_bom_path: Path) -> Optional[dict | None]:
+    """Get dependency components.
+
+    Parameters
+    ----------
+    root_bom_path : str
+        Path to the root bom.json file.
+
+    Returns
+    -------
+    Optional[dict|None]
+        The root component.
+    """
+    try:
+        root_bom = deserialize_bom_json(root_bom_path)
+        return root_bom.get("metadata").get("component")  # type: ignore
+    except (CycloneDXParserError, KeyError) as error:
+        logger.error(error)
+
+    return None
+
+
+def get_dep_components(
+    root_bom_path: Path, child_bom_paths: Optional[list[Path]] = None, recursive: bool = False
+) -> Iterable[dict]:
+    """Get dependency components.
+
+    Parameters
+    ----------
+    root_bom_path : str
+        Path to the root bom.json file.
+    child_bom_paths: Optional[list[Path]]
+        The list of paths to sub-project bom.json files.
+    recursive: bool
+        Set to False to get the direct dependencies only (default).
+
+    Yields
+    ------
+    dict
+        The dependencies as CycloneDX components.
+    """
+    bom_objects: list[dict] = []
+    try:
+        root_bom = deserialize_bom_json(root_bom_path)
+        components = root_bom.get("components")
+        bom_objects.append(root_bom)
+    except (CycloneDXParserError, KeyError) as error:
+        logger.error(error)
+        return
+
+    if components is None:
+        logger.error("The BOM file at %s misses components.", str(root_bom_path))
+        return
+
+    dependencies = []
+    modules = set()  # Stores all module dependencies.
+    for child_path in child_bom_paths or []:
+        try:
+            bom_objects.append(deserialize_bom_json(child_path))
+        except CycloneDXParserError as error:
+            logger.error(error)
+            continue
+    for bom in bom_objects:
+        try:
+            bom_ref = bom.get("metadata").get("component").get("bom-ref")  # type: ignore
+            if bom_ref:
+                modules.add(bom_ref)
+            for node in bom.get("dependencies"):  # type: ignore
+                if node.get("ref") == bom_ref or recursive:
+                    dependencies.extend(node.get("dependsOn"))
+        except KeyError as error:
+            logger.debug(error)
+
+    for dependency in dependencies:
+        if dependency in modules:
```

Comment:
I don't quite understand why do we need to ignore the dependencies that already existed in `modules`.

Reply:
See my answer [here](https://github.com/oracle-samples/macaron/pull/57#discussion_r1107885322).